client/src/main/java/io/pravega/client/stream/impl/EventStreamWriterImpl.java

```diff
@@ -66,30 +68,31 @@
         Preconditions.checkNotNull(serializer);
         this.stream = stream;
         this.controller = controller;
+        this.segmentSealedCallBack = this::handleLogSealed;
```

Comment:
This method is annotated with GuardedBy which implies that it is invalid to call it from any location that does hold the specified lock. Invocations from this member will not honor that contract. You're going to need to modify things so that the lock is held. The easiest way to do that is probably just to grab the lock inside of the method. However the semantics of that are complicated. The point of the lock is to ensure that ordering guarantees hold in the event that there is a concurrent segment sealed and write/close/flush. In such a case we would need to deal with the case where a writer already holds the lock and the seal is blocked on it. probably the easiest thing would be to let the writer encounter the exception and handle it directly and have the handleLogSealed detect that it already performed the operation when it was subsequently. invoked. Either way, work through it and document why the ordering is guaranteed to be correct.